### PR TITLE
Implement `IdentityCreate` state transition

### PR DESCRIPTION
**Pull request by jawid-h**

### CODE REVIEW

File: test/unit/identity/stateTransition/IdentityCreateStateTransition.spec.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  '../../../../lib/stateTransition/stateTransitionTypes',
+);
+
+describe('IdentityCreateStateTransition', () => {
+  let rawStateTransition;
+  let stateTransition;
+
+  beforeEach(() => {
+    rawStateTransition = {
+      identityCreateStateTransitionVersion: 1,
```
Comment (antouhou) [inline test/unit/identity/stateTransition/IdentityCreateStateTransition.spec.js]: I think we need to move this to fixtures

File: lib/identity/stateTransitions/IdentityCreateStateTransition.js [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+    return this;
+  }
+
+  getIdentityStateTransitionVersion() {
+    return this.identityCreateStateTransitionVersion;
+  }
+
+  /**
+   * Sets an outPoint. OutPoint is a poiner to the output funding identity creation.
```
Comment (thephez) [inline lib/identity/stateTransitions/IdentityCreateStateTransition.js]:
``` suggestion
   * Sets an outPoint. OutPoint is a pointer to the output funding identity creation.
```

File: test/integration/identity/stateTransition/validation/structure/validateIdentityCreateSTStructureFactory.spec.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+    expectValidationError(result, JsonSchemaError, 1);
+
+    const [error] = result.getErrors();
+
+    expect(error.keyword).to.equal('minItems');
+    expect(error.dataPath).to.equal('.publicKeys');
+  });
+
+  it('should throw an error if publicKeys have mote than 10 items', () => {
```
Comment (thephez) [inline test/integration/identity/stateTransition/validation/structure/validateIdentityCreateSTStructureFactory.spec.js]:
```suggestion
  it('should throw an error if publicKeys have more than 10 items', () => {
```

File: schema/model/identity.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      "minimum": 0,
+      "maximum": 65535
+    },
+    "publicKeys": {
+      "type": "array",
+      "items": {
+        "$ref": "https://schema.dash.org/dpp-0-4-0/base/public-key"
+      },
+      "minItems": 1,
+      "maxItems": 100
```
Comment (thephez) [inline schema/model/identity.json]: Is this supposed to be 10 or 100? [identity-create.json](https://github.com/dashevo/js-dpp/pull/77/files#diff-3542b91b9a79311ee219fa1351b7d5f4) appears to limit the publicKeys array to size 10, but I may be missing something...

Comment (antouhou) [reply schema/model/identity.json]: It's how many keys you generally allowed to store in your identity, while 10 is how many you can add at once. I don't know if that maks sense, those are just arbitrary numbers

File: lib/errors/DuplicatedIdentitySTPublicKeyError.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  constructor(stateTransition) {
+    super('Duplicated public keys found in the identity state transition');
+
+    this.stateTransition = stateTransition;
+  }
+
+  /**
+   * Get Document
+   *
+   * @return {Document}
```
Comment (shuplenkov) [inline lib/errors/DuplicatedIdentitySTPublicKeyError.js]: Get StateTransition

`{IdentityCreateStateTransition}`

File: lib/errors/DuplicatedIdentitySTPublicKeyIdError.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  constructor(stateTransition) {
+    super('Duplicated public key ids found in the identity state transition');
+
+    this.stateTransition = stateTransition;
+  }
+
+  /**
+   * Get Document
+   *
+   * @return {Document}
```
Comment (shuplenkov) [inline lib/errors/DuplicatedIdentitySTPublicKeyIdError.js]: Get StateTransition

{IdentityCreateStateTransition}

File: lib/errors/IdentityAlreadyExistsError.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+    this.stateTransition = stateTransition;
+  }
+
+  /**
+   * Get Document
+   *
+   * @return {Document}
+   */
+  getStateTransition() {
```
Comment (shuplenkov) [inline lib/errors/IdentityAlreadyExistsError.js]: Get StateTransition

{IdentityCreateStateTransition}

File: lib/errors/IdentitySTWrongVersionError.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+   */
+  constructor(stateTransition, currentVersion) {
+    super('Identity state transition version is too high');
+
+    this.name = this.constructor.name;
+    this.currentVersion = currentVersion;
+
+    this.stateTransition = stateTransition;
+
+    if (Error.captureStackTrace) {
```
Comment (shuplenkov) [inline lib/errors/IdentitySTWrongVersionError.js]: Why do you use  `captureStackTrace` and `this.name`, but not in other such errors?

File: lib/errors/DuplicatedIdentitySTPublicKeyError.js (hunk 5 above)
Comment (antouhou) [reply lib/errors/DuplicatedIdentitySTPublicKeyError.js]: Fixed

File: lib/errors/DuplicatedIdentitySTPublicKeyIdError.js (hunk 6 above)
Comment (antouhou) [reply lib/errors/DuplicatedIdentitySTPublicKeyIdError.js]: Fixed

File: lib/errors/IdentityAlreadyExistsError.js (hunk 7 above)
Comment (antouhou) [reply lib/errors/IdentityAlreadyExistsError.js]: Fixed

File: lib/errors/IdentitySTWrongVersionError.js (hunk 8 above)
Comment (antouhou) [reply lib/errors/IdentitySTWrongVersionError.js]: That's true, that doesn't make any sense since this code is already in ConsensusError

Comment (antouhou) [reply lib/errors/IdentitySTWrongVersionError.js]: Fixed

File: lib/errors/IncorrectIdentityTypeError.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, this.constructor);
+    }
+  }
+
+  /**
+   * Get failed state transition
+   *
+   * @return {IdentityCreateStateTransition}
```
[the remark this answers is not in the document]
Comment (antouhou) [reply lib/errors/IncorrectIdentityTypeError.js]: Fixed

File: lib/identity/errors/InvalidIdentityError.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+   * @return {ConsensusError[]}
+   */
+  getErrors() {
+    return this.errors;
+  }
+
+  /**
+   * Get raw Document
+   *
+   * @return {RawDocument}
```
[the remark this answers is not in the document]
Comment (antouhou) [reply lib/identity/errors/InvalidIdentityError.js]: Fixed

File: lib/identity/model/Identity.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    throw new WrongStateTransitionTypeError(stateTransition);
+  }
+
+  /**
+   * @param {string} id
+   * @return {Identity}
+   */
+  setId(id) {
+    this.id = id;
+    return this;
```
[the remark this answers is not in the document]
Comment (antouhou) [reply lib/identity/model/Identity.js]: Fixed

File: lib/errors/IncorrectIdentityTypeError.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+   * @param {Identity|RawIdentity} identity
+   */
+  constructor(identity) {
+    super('Identity type is within the reserved types range, but is unknown to the protocol');
+
+    this.name = this.constructor.name;
+
+    this.identity = identity;
+
+    if (Error.captureStackTrace) {
```
[the remark this answers is not in the document]
Comment (antouhou) [reply lib/errors/IncorrectIdentityTypeError.js]: Fixed

File: lib/errors/UnknownIdentityTypeError.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+   * @param {IdentityCreateStateTransition} stateTransition
+   */
+  constructor(stateTransition) {
+    super('Identity type is within the reserved types range, but is unknown to the protocol');
+
+    this.name = this.constructor.name;
+
+    this.stateTransition = stateTransition;
+
+    if (Error.captureStackTrace) {
```
[the remark this answers is not in the document]
Comment (antouhou) [reply lib/errors/UnknownIdentityTypeError.js]: Fixed

File: lib/identity/Identity.js [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+   */
+  getPublicKeys() {
+    return this.publicKeys;
+  }
+
+  /**
+   * Returns a public key for a given id
+   *
+   * @param {number} keyId
+   * @return {RawIdentityPublicKey}
```
Comment (shuplenkov) [inline lib/identity/Identity.js]: It seems to be 
`@return {IdentityPublicKey}`

File: lib/identity/Identity.js [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+Identity.TYPES = {
+  USER: 1,
+  APPLICATION: 2,
+};
+
+Identity.MAX_RESERVED_TYPE = 32767;
+
+Identity.TYPES_ENUM = Object
+  .keys(Identity.TYPES)
+  .map(typeName => Identity.TYPES[typeName]);
```
Comment (shuplenkov) [inline lib/identity/Identity.js]: Why not just 
`Identity.TYPES_ENUM = Object.values(Identity.TYPES);`

File: test/unit/identity/Identity.spec.js [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+  let encodeMock;
+
+  beforeEach(function beforeEach() {
+    rawIdentity = {
+      id: 'someId',
+      type: 0,
+      publicKeys: [
+        {
+          id: 1,
+          type: 1,
```
Comment (shuplenkov) [inline test/unit/identity/Identity.spec.js]: `type: IdentityPublicKey.TYPES.ECDSA_SECP256K`

File: test/unit/identity/IdentityPublicKey.spec.js [hunk 17]
```diff
@@ -0,0 +1,103 @@
+const IdentityPublicKey = require('../../../lib/identity/IdentityPublicKey');
+
+describe('IdentityPublicKey', () => {
+  let rawPublicKey;
+  let publicKey;
+
+  beforeEach(() => {
+    rawPublicKey = {
+      id: 1,
+      type: 1,
```
Comment (shuplenkov) [inline test/unit/identity/IdentityPublicKey.spec.js]: `type: IdentityPublicKey.TYPES.ECDSA_SECP256K`

File: test/unit/identity/stateTransitions/identityCreateTransition/IdentityCreateTransition.spec.js [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+  let IdentityCreateTransition;
+
+  beforeEach(function beforeEach() {
+    rawStateTransition = {
+      lockedOutPoint: 'c3BlY2lhbEJ1ZmZlcg==',
+      identityType: 1,
+      publicKeys: [
+        {
+          id: 1,
+          type: 1,
```
Comment (shuplenkov) [inline test/unit/identity/stateTransitions/identityCreateTransition/IdentityCreateTransition.spec.js]: `type: IdentityPublicKey.TYPES.ECDSA_SECP256K`

File: lib/identity/validation/validateIdentityType.js [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * @typedef validateIdentityType
+ * @param {number} type
+ * @return ValidationResult
+ */
+function validateIdentityType(type) {
+  const result = new ValidationResult();
+
+  const isReservedType = type < Identity.MAX_RESERVED_TYPE;
+  const isRegisteredType = Object.values(Identity.TYPES).includes(type);
```
Comment (antouhou) [inline lib/identity/validation/validateIdentityType.js]: There's a constant in the Identity class for this so we don't need to do this every time, `Identity.TYPES_ENUM`

File: test/unit/stateTransition/validation/validateIdentityExistenceAndTypeFactory.spec.js [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+  let rawIdentityUser;
+
+  beforeEach(function beforeEach() {
+    dataProviderMock = createDataProviderMock(this.sinonSandbox);
+
+    validateIdentityExistenceAndType = validateIdentityExistenceAndTypeFactory(
+      dataProviderMock,
+    );
+
+    userId = 'iTYF+bWBA4MYRURcsBpBkgfwiqV7sYVnTDPR4uQ/KLU=';
```
Comment (antouhou) [inline test/unit/stateTransition/validation/validateIdentityExistenceAndTypeFactory.spec.js]: Need to change to base58, as we agreed that this is the best way to make browser links in explorers work

**Pull request merged**